src/Symfony/Component/HttpFoundation/JsonResponse.php

```diff
@@ -194,6 +196,18 @@ protected function update()
             $this->headers->set('Content-Type', 'application/json');
         }
 
-        return $this->setContent($this->data);
+        return $this->setContent($serializedData);
+    }
+
+    /**
+     * @param mixed $content
+     *
+     * @return Response
+     */
+    public function setContent($content)
+    {
+        $this->data = json_decode($content);
```

Comment:
this is broken in case the content is not a JSON string (which is the case for JSONP cases in your code currently btw)